File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>
#include <queue>
#include <utility>
#include <iostream>
using namespace std;

// 현재 위치에 도달하는 방법: 이전의 위치에서 오른쪽 이동 or 이전 위치에서 아래쪽 이동 

// dp[n][m] = ans 

int divisor = 1000000007;

int solution(int m, int n, vector<vector<int>> puddles) {
    vector<vector<int>> dp(n+1, vector<int>(m+1, 0));
    // 초기화
    dp[1][1] = 1;
    for(int i=0; i<puddles.size(); i++){
        dp[puddles[i][1]][puddles[i][0]] = -2;
    }
    
    for(int i=1; i<=n; i++){
        for(int j=1; j<=m; j++){
            if(dp[i][j] == -2) continue;
            if(dp[i-1][j] != -2){
                dp[i][j] += dp[i-1][j];
            }
            if(dp[i][j-1] != -2){
                dp[i][j] += dp[i][j-1];
            }
            dp[i][j] %= divisor;
        }
    }
    
    int answer = dp[n][m];
    return answer;
}

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;

// java modular 연산 음수 가능한거 인지 
// (1,1) -> (n,m)
class Solution {
    int divisor = 1000000007;
    int[][] memo;
    int max_y, max_x;
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        memo = new int[n+1][m+1];
        max_y = n;
        max_x = m;
        // init -1, puddles 0, 
        for(int[] x: memo){
            for(int i=0; i<x.length; i++){
                x[i] = -1;
            }
        }
        memo[1][1] = 1; // 출발지점 1가지 
        for(int[] puddle: puddles){
            memo[puddle[1]][puddle[0]] = 0; // puddle 0가지 
        }
        answer = getMem(n, m);
        return answer;
    }
    
    // y,x 좌표 , memo[y][x] = memo[y-1][x] + memo[y][x-1] 
    // x좌표 증가 or y좌표 증가로만 이동함 
    public int getMem(int y, int x){
        if(y <= 0 || y > max_y || x <= 0 || x > max_x){
            return 0;
        }
        if(memo[y][x] != -1){
            return memo[y][x];
        }
        
        int tot = (getMem(y-1, x) % divisor) + (getMem(y,x-1) % divisor);
        tot %= divisor;
        memo[y][x] = tot;
        return memo[y][x];
    }
}